### PR TITLE
[AscendNPU-IR][Expert] Bug fix for DSv4 attn high-pref kernel

### PR DESCRIPTION
**Pull request by voidLitchi**
### Change Summary  
This commit addresses multiple issues in the sparse attention operator, including precision problems, device compatibility, and compilation errors, while also optimizing the test data generation algorithm.

### Detailed Changes  

1. **examples/deepseek_v4/example_sparse_attn_kernel_highperf.py**  

   - **Precision fix (L62)**: Fixed a precision issue when `num_top_k_blocks` evaluates to 1.  
     - Before: `(top_k - 1) // block + 1` could cause precision loss under specific boundary conditions.  
     - After: Ensures exact computation when the result is 1.  

   - **Device generalization improvement (L695)**: Dynamically obtains the number of AI cores using `NPUUtils.get().get_aicore_num()`.  
     - Enhances generalization across different NPU devices by avoiding hardcoded device parameters.  

   - **Test data generation standardization (L804-807)**: Standardized the generation algorithm for `top_k_indices` in causal mode.  
     - Uses a more reasonable random number generation method to ensure test data complies with the calling context in the DSV4 model.  

2. **src/op/ascend.cc**  

   - **Compilation error fix (L35)**: Resolved a compilation error that occurred when `top_k` was not divisible by the block size.  
     - Treats the TIR expression as a valid scalar to prevent compilation errors.

### CODE REVIEW

File: examples/deepseek_v4/example_sparse_attn_kernel_highperf.py [hunk 1]
```diff
@@ -780,23 +792,27 @@ def rand_sparse_attn_input(
...
     if causal:
-        # Apply causal mask on top_k_indices
-        max_len = max(seq_len, top_k)
-        causal_mask = torch.tril(torch.ones(max_len, max_len)).to(top_k_indices.device)
-        causal_mask = causal_mask[:seq_len, :top_k]
-        causal_mask = causal_mask.unsqueeze(dim=0).bool()
-        top_k_indices = torch.where(causal_mask, top_k_indices, -1)
+        rand = torch.rand(batch_size, seq_len, top_k, device=device)
+        upper = torch.arange(1, seq_len + 1, device=device).view(1, seq_len, 1)
+        top_k_indices = (rand * upper).to(torch.int32)
```
[the remark this answers is not in the document]
Comment (voidLitchi) [reply examples/deepseek_v4/example_sparse_attn_kernel_highperf.py]: Fixed

File: src/op/ascend.cc [hunk 2]
```diff
@@ -32,14 +32,7 @@ NpuirOperand NpuirOperand::FromExpr(const PrimExpr &expr,
...
   }
-  if (expr.as<IntImm>() || expr.as<FloatImm>() || expr.as<tir::VarNode>()) {
-    // If there are other types of nodes that need to be treated as scalars,
-    // please add them here.
-    return NpuirOperand::Scalar(expr);
-  }
-  LOG(FATAL) << "NpuirOperand::FromExpr cannot handle the expr with type of \""
-             << expr->GetTypeKey() << '\"';
-  __builtin_unreachable();
+  return NpuirOperand::Scalar(expr);
```
Comment (CeleNewYear) [inline src/op/ascend.cc]: Why was the exception catching and handling removed? This will cause illegal data types to be processed as scalars, making problem diagnosis more difficult and increasing risk.

**Review by CeleNewYear — CHANGES_REQUESTED** [review]

File: src/op/ascend.cc (hunk 2 above)
Comment (voidLitchi) [reply src/op/ascend.cc]: I think you're right. I restored the original restrictions but added some arithmetic expression nodes to the conditional statements. However, this might still cause issues when dealing with more scenarios. Later, we could add more allowed node types as needed, or consider refactoring the logic here.

**Review by CeleNewYear — APPROVED** [review]
LGTM
